src/state-visualizer/visualizer.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+import graphviz
+import time
+from PyQt5.QtWidgets import *
+from PyQt5.QtCore import *
+from PyQt5.QtGui import QPainter
+from PyQt5.QtSvg import QSvgRenderer
+
+import rospy
+from smach_msgs.msg import SmachContainerStatus, SmachContainerStructure
+from threading import Lock
+from dataclasses import dataclass
+from typing import Optional, List, Dict
+
+
+@dataclass
+class State:
+    name: str
+    children: List[State]
+
+
+class StateMachine:
+    def __init__(self):
+        self.states: Dict[str, State] = {}
+        self.structure: Optional[SmachContainerStructure] = None
+        self.mutex: Lock = Lock()
+        self.cur_active: str = ""
+        self.needs_redraw: bool = True
+
+    def set_active_state(self, active_state):
+        """
+        sets the state specified to be active
+        """
+        if active_state != self.cur_active and active_state in self.states:
+            self.cur_active = active_state
+            self.needs_redraw = True
+
+    def rebuild(self, structure: SmachContainerStructure):
+        """
+        rebuilds the state dictionary with a new structure message
+        """
+        self.states = {child: State(child, []) for child in structure.children}
+        for start, end in zip(structure.outcomes_from, structure.outcomes_to):
+            if end != "None":
+                self.states[start].children.append(self.states[end])
+        self.needs_redraw = True
+
+    def check_rebuild(self, structure: SmachContainerStructure):
+        """
+        checks if the structure passed as input matches the structure already represented
+        """
+        if structure == self.structure:
+            return False
+        else:
+            self.rebuild(structure)
+
+
+state_machine = StateMachine()
```

Comment:
lets avoid global scope, this can get passed by ref to the constructor of the GUI